### PR TITLE
Fixed #27226 -- Removed Last-Modified header setting in cache.patch_response_headers().

### PR DESCRIPTION
**Pull request by renskiy**

### CODE REVIEW

**Comment by timgraham** [comment]
This is backwards-incompatible isn't it? Perhaps a deprecation similar to what's proposed in https://github.com/django/django/pull/6393 would solve the issue.

**Comment by renskiy** [comment]
Do you propose to deprecate entire patch_response_headers()? And add another function instead?

**Comment by timgraham** [comment]
https://github.com/django/django/pull/6393 deprecates the `USE_ETAGS` branch in `patch_response_headers()`.

As for the `'Last-Modified'` branch, it's not clear to me what problem it causes. Now these headers might disappear wherever the function is used? The ticket needs more explanation about the problems this solves and a consideration of backwards compatibility.

**Comment by renskiy** [comment]
Ok with `USE_ETAGS`.

About `Last-Modified` - it is simply not needed. Nobody who uses `patch_response_headers` does check `If-Modified-Since`/`If-Unmodified-Since` from requests then.

**Comment by timgraham** [comment]
I haven't really used these headers, but I don't understand what you're saying. How can you know which headers a client is looking at?

**Comment by renskiy** [comment]
There is [RFC7234](https://tools.ietf.org/html/rfc7234) about how clients should treat HTTP-caching headers. [Section 4.2.2](https://tools.ietf.org/html/rfc7234#section-4.2.2) says:
> A cache MUST NOT use heuristics to determine freshness when an explicit expiration time is present in the stored response.

In other words there is no reason to use `Last-Modified` among with valid `Expires`

**Comment by timgraham** [comment]
Does the header cause some harm though?

Anyway, it might be safest to proceed with the deprecation in https://github.com/django/django/pull/6393 and then remove Last-Modified when that deprecation completes. Would you like to try updating that patch and addressing the comments?

**Comment by timgraham** [comment]
I updated the USE_ETAGS deprecation patch in https://github.com/django/django/pull/7256. Feel free to review.

As for `Last-Modified`, it's been in Django since 1.0 (8aa98af6). I don't know if removing it will cause any harm or benefit so I guess I'm inclined to leave it unless you have a strong rationale for removing it. I can't think of a practical way to deprecate it, so I think we'd just have to remove it outright and call it a backwards-incompatible change.

**Comment by renskiy** [comment]
There is one case where `Last-Modified` may lead to implicit caching of response by browsers (using special algorithm based on difference between date of `Last-Modified` and current date). But this works only if there is no any explicit HTTP-caching header in response: `Cache-Control` or `Expires`). And because `patch_response_headers` sets both of these headers there is no any benefit from `Last-Modified`. Also there is no harm if `Last-Modified` will be removed from `patch_response_headers `.

We can remove it now or we can remove it along with `USE _ETAGS` (#7256).

**Comment by timgraham** [comment]
I didn't do a careful analysis, but could there be cases where Django makes use of Last-Modified such that if it's no longer set, could cause a behavior changed? e.g. https://github.com/django/django/blob/b5aac66b28c615b2bda63548cbd695dbb5a0c381/django/middleware/http.py#L20-L22

**Comment by renskiy** [comment]
Clients use `If-Modified-Since` and `If-Unmodified-Since` in requests if there was `Last-Modified` header in previous answer from server. There are only two places where `HTTP_IF_MODIFIED_SINCE`/`HTTP_IF_UNMODIFIED_SINCE` are checked:

1. `ConditionalGetMiddleware`
2. `last_modified` decorator

And nothing in other place, including `UpdateCacheMiddleware`/`FetchFromCacheMiddleware`

**Comment by timgraham** [comment]
I was thinking about a case where you have:
```
MIDDLWARE_CLASES = (
    'django.middleware.http.ConditionalGetMiddleware',
    'django.middleware.cache.UpdateCacheMiddleware',
    ...,
)
```
If `patch_response_headers()` (called by `UpdateCacheMiddleware`) no longer sets `Last-Modified`, then a `last_modified` value won't be passed to `ConditionalGetMiddleware`'s`get_conditional_response()`. I'm not sure if that could cause a problem or behavior difference in practice.

A found a few projects using GitHub code search that have the middleware ordering above although more often the order is swapped in which case I don't think there would be a behavior change.

**Comment by renskiy** [comment]
But there is no way when `Last-Modified` generated from `UpdateCacheMiddleware` will take effect in `ConditionalGetMiddleware` in such case. Because `UpdateCacheMiddleware` always generates fresh one and `ConditionalGetMiddleware` does always miss (`If-Modified-Since` always less then `Last-Modified`).

**Comment by timgraham** [comment]
Okay, can you update patch to remove only Last-Modified and also add a mention and explanation of the reason for the removal in the 1.11 release notes under "Backwards-incompatible changes"?

File: django/utils/cache.py [hunk 1]
```diff
@@ -211,14 +210,7 @@ def patch_response_headers(response, cache_timeout=None):
...
         cache_timeout = 0  # Can't have max-age negative
-    if settings.USE_ETAGS and not response.has_header('ETag'):
-        if hasattr(response, 'render') and callable(response.render):
-            response.add_post_render_callback(set_response_etag)
-        else:
-            response = set_response_etag(response)
-    if not response.has_header('Last-Modified'):
-        response['Last-Modified'] = http_date()
-    if not response.has_header('Expires'):
+    if 'Expires' not in response:
```
Comment (timgraham) [inline django/utils/cache.py]: Don't make an unrelated cosmetic change here.

**Comment by renskiy** [comment]
ok, I'll fix it

**Comment by renskiy** [comment]
ready to check

**Comment by timgraham** [comment]
merged in 9e07a9b5feb30fe06fb23d0e77edd34f6c3f92b5, thanks!
